File: static/js/marketplace.js
document.addEventListener('DOMContentLoaded', () => {
    // Modal Elements
    const registerModal = document.getElementById('registerModal');
    const sellModal = document.getElementById('sellModal');
    const buyModal = document.getElementById('buyModal');

    // Button Elements
    const registerBtn = document.getElementById('registerCompanyBtn');
    const sellBtn = document.getElementById('sellCreditsBtn');
    const buyBtn = document.getElementById('buyCreditsBtn');
    const refreshListingsBtn = document.getElementById('refreshListingsBtn');
    const refreshTransactionsBtn = document.getElementById('refreshTransactionsBtn');
    const applyFiltersBtn = document.getElementById('applyFiltersBtn');
    const tradersModal = document.getElementById('tradersModal');
    const activeTradersCard = document.getElementById('activeTradersCard');

    // Close buttons
    const closeButtons = document.querySelectorAll('.close-modal');

    // Open modals
    registerBtn.addEventListener('click', () => openModal(registerModal));
    sellBtn.addEventListener('click', () => openModal(sellModal));
    buyBtn.addEventListener('click', () => openModal(buyModal));
    if (activeTradersCard) {
        activeTradersCard.addEventListener('click', () => {
            openModal(tradersModal);
            loadActiveTraders();
        });
    }

    // Close modals
    closeButtons.forEach(btn => {
        btn.addEventListener('click', () => {
            const modalId = btn.getAttribute('data-modal');
            closeModal(document.getElementById(modalId));
        });
    });

    // Close on outside click
    window.addEventListener('click', (e) => {
        if (e.target.classList.contains('modal')) {
            closeModal(e.target);
        }
    });

    // Form submissions
    document.getElementById('registerForm').addEventListener('submit', handleRegisterCompany);
    document.getElementById('sellForm').addEventListener('submit', handleCreateListing);
    document.getElementById('buyForm').addEventListener('submit', handleBuyOrder);

    // Refresh buttons
    refreshListingsBtn.addEventListener('click', loadMarketplaceListings);
    refreshTransactionsBtn.addEventListener('click', loadTransactions);
    applyFiltersBtn.addEventListener('click', applyFilters);

    // ========== FUNCTIONS (Exposed Globally for SPA) ==========

    window.loadMarketplaceData = () => {
        loadMarketStats();
        loadMarketplaceListings();
        loadTransactions();
        loadActiveTraders(); // Keep traders list warm
    };

    // Initial load
    window.loadMarketplaceData();

    // Auto-refresh every 30 seconds
    setInterval(() => {
        window.loadMarketplaceData();
    }, 30000);

    // Internal functions remaining...

    function openModal(modal) {
        modal.classList.remove('hidden');
    }

    function closeModal(modal) {
        modal.classList.add('hidden');
        // Clear result messages
        modal.querySelectorAll('.result-message').forEach(msg => {
            msg.classList.add('hidden');
            msg.textContent = '';
        });
    }

    async function loadMarketStats() {
        try {
            const response = await fetch('/api/marketplace/stats');
            const data = await response.json();

            if (data.success) {
                const stats = data.stats;
                document.getElementById('totalVolume').textContent = `$${stats.total_volume_usd.toLocaleString()}`;
                document.getElementById('activeListings').textContent = stats.active_listings;
                document.getElementById('totalTrades').textContent = stats.total_transactions;
                document.getElementById('avgPrice').textContent = `$${stats.average_price_per_credit.toFixed(2)}`;
                document.getElementById('activeTraders').textContent = stats.total_companies;
            }
        } catch (error) {
            console.error('Error loading market stats:', error);
        }
    }

    async function loadMarketplaceListings(filters = null) {
        try {
            let url = '/api/marketplace/listings';
            if (filters) {
                const params = new URLSearchParams(filters);
                url += `?${params.toString()}`;
            }

            const response = await fetch(url);
            const data = await response.json();

            const grid = document.getElementById('listingsGrid');

            if (!data.success || data.listings.length === 0) {
                grid.innerHTML = `
                    <div class="empty-state-marketplace">
                        <div class="empty-icon">📦</div>
                        <p>No active listings found. Try adjusting your filters or be the first to sell!</p>
                    </div>
                `;
                return;
            }

            grid.innerHTML = '';

            data.listings.forEach(listing => {
                const card = createListingCard(listing);
                grid.appendChild(card);
            });

        } catch (error) {
            console.error('Error loading listings:', error);
        }
    }

    function createListingCard(listing) {
        const card = document.createElement('div');
        card.className = 'listing-card';

        const createdDate = new Date(listing.created_at * 1000).toLocaleDateString();

        card.innerHTML = `
            <div class="listing-header">
                <div class="listing-seller">${listing.seller_name}</div>
                <div class="listing-badge">VERIFIED SELLER</div>
            </div>
            
            <div class="listing-details">
                <div class="listing-detail-row">
                    <span class="listing-detail-label">Asset:</span>
                    <span class="listing-detail-value">Verified Carbon Credit</span>
                </div>
                <div class="listing-detail-row">
                    <span class="listing-detail-label">Quantity:</span>
                    <span class="listing-detail-value">${listing.available_amount} tCO2e</span>
                </div>
                <div class="listing-detail-row">
                    <span class="listing-detail-label">Price:</span>
                    <span class="listing-detail-value listing-price">$${listing.price_per_credit.toFixed(2)}</span>
                </div>
                <div class="listing-detail-row">
                    <span class="listing-detail-label">Location:</span>
                    <span class="listing-detail-value">${listing.location}</span>
                </div>
                ${listing.description ? `
                <div style="margin-top: 1rem; padding-top: 0.5rem; border-top: 1px dashed var(--border);">
                    <p style="font-size: 0.8rem; color: var(--text-muted); margin: 0;">${listing.description}</p>
                </div>` : ''}
            </div>
            
            <div class="listing-actions">
                <button class="btn-buy" onclick="sendInquiry('${listing.listing_id}', '${listing.seller_name}')">
                    Send Inquiry
                </button>
                <button class="btn-details" onclick="viewListingDetails('${listing.listing_id}')">
                    Full Details
                </button>
            </div>
        `;

        return card;
    }

    async function loadTransactions() {
        try {
            const response = await fetch('/api/marketplace/transactions');
            const data = await response.json();

            const list = document.getElementById('transactionsList');

            if (!data.success || data.transactions.length === 0) {
                list.innerHTML = `
                    <div class="empty-state-marketplace">
                        <div class="empty-icon">📋</div>
                        <p>No transactions yet. Start trading to see your history!</p>
                    </div>
                `;
                return;
            }

            list.innerHTML = '';

            // Show last 10 transactions
            data.transactions.slice(-10).reverse().forEach(txn => {
                const item = createTransactionItem(txn);
                list.appendChild(item);
            });

        } catch (error) {
            console.error('Error loading transactions:', error);
        }
    }

    function createTransactionItem(txn) {
        const item = document.createElement('div');
        item.className = 'transaction-item';

        const date = new Date(txn.timestamp * 1000).toLocaleString();

        item.innerHTML = `
            <div class="transaction-icon">✅</div>
            <div class="transaction-details">
                <div class="transaction-title">
                    ${txn.buyer_name} ← ${txn.seller_name}
                </div>
                <div class="transaction-meta">
                    ${txn.transaction_id} | ${date}
                </div>
            </div>
            <div class="transaction-amount">
                <div class="transaction-value">$${txn.total_price.toLocaleString()}</div>
                <div class="transaction-credits">${txn.credit_amount} tCO2e @ $${txn.price_per_credit.toFixed(2)}</div>
            </div>
        `;

        return item;
    }

    async function loadActiveTraders() {
        const tradersList = document.getElementById('tradersList');
        if (!tradersList) return;

        try {
            const response = await fetch('/api/marketplace/companies');
            const data = await response.json();

            if (!data.success || data.companies.length === 0) {
                tradersList.innerHTML = '<div class="empty-state">No active enterprises found.</div>';
                return;
            }

            tradersList.innerHTML = '';
            data.companies.forEach(company => {
                const card = document.createElement('div');
                card.className = 'trader-card';
                card.innerHTML = `
                    <div class="trader-badge">VERIFIED</div>
                    <div class="trader-logo">🏢</div>
                    <div class="trader-info">
                        <h3>${company.name}</h3>
                        <p>${company.industry} | ${company.country}</p>
                        <p style="font-size: 0.75rem; color: var(--accent); margin-top: 0.2rem;">${company.email}</p>
                    </div>
                    <div class="trader-stats">
                        <div class="t-stat">
                            <span class="t-label">Credits Owned</span>
                            <span class="t-value">${company.credits_owned.toFixed(2)}</span>
                        </div>
                        <div class="t-stat">
                            <span class="t-label">Total Trades</span>
                            <span class="t-value">${company.total_trades}</span>
                        </div>
                    </div>
                    <button class="btn-secondary" style="margin-top: 1rem; width: 100%; font-size: 0.8rem;" onclick="viewCompanyProfile('${company.company_id}')">
                        View Profile
                    </button>
                `;
                tradersList.appendChild(card);
            });
        } catch (error) {
            console.error('Error loading traders:', error);
            tradersList.innerHTML = '<div class="error-state">Failed to load trader data</div>';
        }
    }

    window.viewCompanyProfile = function (companyId) {
        alert("Detailed Company Profiles coming soon! ID: " + companyId);
    };

    function applyFilters() {
        const maxPrice = document.getElementById('maxPriceFilter').value;
        const minAmount = document.getElementById('minAmountFilter').value;

        const filters = {};
        if (maxPrice) filters.max_price = maxPrice;
        if (minAmount) filters.min_amount = minAmount;

        loadMarketplaceListings(filters);
    }

    async function handleRegisterCompany(e) {
        e.preventDefault();

        const resultDiv = document.getElementById('registrationResult');
        resultDiv.classList.add('hidden');

        const formData = {
            company_name: document.getElementById('companyName').value,
            industry: document.getElementById('industry').value,
            country: document.getElementById('country').value,
            email: document.getElementById('companyEmail').value
        };

        try {
            const response = await fetch('/api/marketplace/register', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(formData)
            });

            const data = await response.json();

            if (data.success) {
                resultDiv.className = 'result-message success';
                resultDiv.textContent = `✅ Success! Your Company ID: ${data.company_id}`;
                resultDiv.classList.remove('hidden');

                // Clear form
                e.target.reset();

                // Update stats
                setTimeout(() => loadMarketStats(), 1000);
            } else {
                throw new Error(data.error || 'Registration failed');
            }
        } catch (error) {
            resultDiv.className = 'result-message error';
            resultDiv.textContent = `❌ Error: ${error.message}`;
            resultDiv.classList.remove('hidden');
        }
    }

    async function handleCreateListing(e) {
        e.preventDefault();

        const resultDiv = document.getElementById('sellResult');
        resultDiv.classList.add('hidden');

        const formData = {
            seller_id: document.getElementById('sellerCompanyId').value,
            credit_amount: parseFloat(document.getElementById('creditAmount').value),
            price_per_credit: parseFloat(document.getElementById('pricePerCredit').value),
            location: document.getElementById('creditLocation').value || 'Unknown',
            description: document.getElementById('creditDescription').value,
            verification_data: {
                verified: true,
                source: 'GeoVerify Sentinel'
            }
        };

        try {
            const response = await fetch('/api/marketplace/create-listing', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(formData)
            });

            const data = await response.json();

            if (data.success) {
                resultDiv.className = 'result-message success';
                resultDiv.textContent = `✅ Listing created! ID: ${data.listing.listing_id}`;
                resultDiv.classList.remove('hidden');

                // Clear form
                e.target.reset();

                // Refresh listings
                setTimeout(() => {
                    loadMarketplaceListings();
                    loadMarketStats();
                }, 1000);
            } else {
                throw new Error(data.error || 'Failed to create listing');
            }
        } catch (error) {
            resultDiv.className = 'result-message error';
            resultDiv.textContent = `❌ Error: ${error.message}`;
            resultDiv.classList.remove('hidden');
        }
    }

    async function handleBuyOrder(e) {
        e.preventDefault();

        const resultDiv = document.getElementById('buyResult');
        resultDiv.classList.add('hidden');

        const formData = {
            buyer_id: document.getElementById('buyerCompanyId').value,
            credit_amount: parseFloat(document.getElementById('buyAmount').value),
            max_price_per_credit: parseFloat(document.getElementById('maxPrice').value)
        };

        try {
            const response = await fetch('/api/marketplace/buy', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(formData)
            });

            const data = await response.json();

            if (data.success) {
                if (data.match && data.match.matched) {
                    resultDiv.className = 'result-message success';
                    resultDiv.textContent = `✅ Order matched! Transaction ID: ${data.match.transaction.transaction_id}`;
                } else {
                    resultDiv.className = 'result-message info';
                    resultDiv.textContent = `ℹ️ Order placed. ${data.match.reason}`;
                }
                resultDiv.classList.remove('hidden');

                // Clear form
                e.target.reset();

                // Refresh data
                setTimeout(() => {
                    loadMarketplaceListings();
                    loadTransactions();
                    loadMarketStats();
                }, 1000);
            } else {
                throw new Error(data.error || 'Failed to create buy order');
            }
        } catch (error) {
            resultDiv.className = 'result-message error';
            resultDiv.textContent = `❌ Error: ${error.message}`;
            resultDiv.classList.remove('hidden');
        }
    }

    // Global functions for inline onclick handlers
    window.sendInquiry = function (listingId, sellerName) {
        const buyerCompanyId = prompt(`Enter your Company ID to send inquiry to ${sellerName}:`);
        if (!buyerCompanyId) return;

        fetch('/api/marketplace/inquiry', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                listing_id: listingId,
                buyer_id: buyerCompanyId
            })
        })
            .then(res => res.json())
            .then(data => {
                if (data.success) {
                    alert(`✅ Inquiry successfully sent to ${sellerName} (${data.seller_email})!\n\nThey have been notified that you are interested in their product.`);
                    loadMarketplaceListings();
                } else {
                    alert(`❌ Error: ${data.error}`);
                }
            })
            .catch(error => {
                alert(`❌ Error: ${error.message}`);
            });
    };

    window.quickBuy = function (listingId, amount, price) {
        const buyerCompanyId = prompt('Enter your Company ID to purchase:');
        if (!buyerCompanyId) return;

        fetch('/api/marketplace/buy', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                buyer_id: buyerCompanyId,
                credit_amount: amount,
                max_price_per_credit: price
            })
        })
            .then(res => res.json())
            .then(data => {
                if (data.success && data.match && data.match.matched) {
                    alert(`✅ Purchase successful! Transaction ID: ${data.match.transaction.transaction_id}`);
                    loadMarketplaceListings();
                    loadTransactions();
                    loadMarketStats();
                } else {
                    alert(`ℹ️ ${data.match ? data.match.reason : 'Order placed'}`);
                }
            })
            .catch(error => {
                alert(`❌ Error: ${error.message}`);
            });
    };

    window.viewListingDetails = function (listingId) {
        alert(`Viewing details for listing: ${listingId}\n\nFull details feature coming soon!`);
    };
});

File: core/marketplace.py
import time
import random
import hashlib
from datetime import datetime

class CarbonMarketplace:
    """
    B2B Carbon Credit Trading Platform for Manufacturing Industry
    Enables verified carbon credits to be listed, bought, and sold on blockchain
    """
    
    def __init__(self, blockchain):
        self.blockchain = blockchain
        self.listings = []  # Active marketplace listings
        self.orders = []    # Buy/Sell orders
        self.transactions = []  # Completed transactions
        self.companies = {}  # Registered companies
        self.listing_id_counter = 1000
        self.order_id_counter = 2000
        self.transaction_id_counter = 3000
        
    def register_company(self, company_name, industry, country, wallet_address, email=None):
        """Register a manufacturing company on the platform"""
        company_id = hashlib.sha256(f"{company_name}{time.time()}".encode()).hexdigest()[:12]
        
        self.companies[company_id] = {
            'company_id': company_id,
            'name': company_name,
            'industry': industry,
            'country': country,
            'email': email or f"contact@{company_name.lower().replace(' ', '')}.com",
            'wallet_address': wallet_address,
            'verified': True,
            'credits_owned': 0,
            'credits_sold': 0,
            'total_trades': 0,
            'reputation_score': 100,
            'joined_date': time.time()
        }
        
        return company_id

    def send_inquiry(self, listing_id, buyer_id):
        """Simulate sending an inquiry email to the seller"""
        listing = next((l for l in self.listings if l['listing_id'] == listing_id), None)
        if not listing:
            return {"success": False, "error": "Listing not found"}
            
        seller = self.companies.get(listing['seller_id'])
        buyer = self.companies.get(buyer_id)
        
        if not seller or not buyer:
            return {"success": False, "error": "Seller or Buyer not found"}
            
        listing['interested_buyers'] += 1
        
        # Simulate email notification
        print(f"NOTIFICATION: Sending email to {seller['email']}")
        print(f"Subject: New Inquiry for your Carbon Credit Listing {listing_id}")
        print(f"Message: Hello {seller['name']}, a buyer ({buyer['name']}) is interested in your listing in {listing['location']}.")
        
        return {
            "success": True, 
            "message": f"Inquiry sent to {seller['name']}",
            "seller_email": seller['email']
        }
    
    def create_listing(self, seller_id, credit_amount, price_per_credit, 
                       verification_data, location, description=""):
        """Create a new carbon credit listing for sale"""
        
        if seller_id not in self.companies:
            return {"success": False, "error": "Company not registered"}
        
        listing_id = f"LST-{self.listing_id_counter}"
        self.listing_id_counter += 1
        
        listing = {
            'listing_id': listing_id,
            'seller_id': seller_id,
            'seller_name': self.companies[seller_id]['name'],
            'credit_amount': credit_amount,
            'available_amount': credit_amount,
            'price_per_credit': price_per_credit,
            'total_value': credit_amount * price_per_credit,
            'verification_data': verification_data,
            'location': location,
            'description': description,
            'status': 'ACTIVE',
            'created_at': time.time(),
            'expires_at': time.time() + (30 * 24 * 60 * 60),  # 30 days
            'views': 0,
            'interested_buyers': 0
        }
        
        self.listings.append(listing)
        
        # Record on blockchain
        block_data = {
            'type': 'LISTING_CREATED',
            'listing_id': listing_id,
            'seller': self.companies[seller_id]['name'],
            'amount': credit_amount,
            'price': price_per_credit,
            'timestamp': time.time()
        }
        
        return {
            'success': True,
            'listing': listing,
            'message': f'Listing {listing_id} created successfully'
        }
    
    def create_buy_order(self, buyer_id, credit_amount, max_price_per_credit):
        """Create a buy order for carbon credits"""
        
        if buyer_id not in self.companies:
            return {"success": False, "error": "Company not registered"}
        
        order_id = f"ORD-{self.order_id_counter}"
        self.order_id_counter += 1
        
        order = {
            'order_id': order_id,
            'buyer_id': buyer_id,
            'buyer_name': self.companies[buyer_id]['name'],
            'credit_amount': credit_amount,
            'max_price_per_credit': max_price_per_credit,
            'status': 'PENDING',
            'created_at': time.time(),
            'matched_listing': None
        }
        
        self.orders.append(order)
        
        # Try to match with existing listings
        match_result = self._match_order(order)
        
        return {
            'success': True,
            'order': order,
            'match': match_result
        }
    
    def _match_order(self, order):
        """Automatically match buy orders with suitable listings"""
        
        # Find listings that match the criteria
        suitable_listings = [
            listing for listing in self.listings
            if listing['status'] == 'ACTIVE'
            and listing['available_amount'] >= order['credit_amount']
            and listing['price_per_credit'] <= order['max_price_per_credit']
        ]
        
        if not suitable_listings:
            return {'matched': False, 'reason': 'No suitable listings found'}
        
        # Sort by best price
        suitable_listings.sort(key=lambda x: x['price_per_credit'])
        best_listing = suitable_listings[0]
        
        # Execute transaction
        transaction = self.execute_transaction(
            order['buyer_id'],
            best_listing['seller_id'],
            best_listing['listing_id'],
            order['credit_amount'],
            best_listing['price_per_credit']
        )
        
        return {
            'matched': True,
            'listing': best_listing,
            'transaction': transaction
        }
    
    def execute_transaction(self, buyer_id, seller_id, listing_id, 
                           credit_amount, price_per_credit):
        """Execute a carbon credit transaction"""
        
        transaction_id = f"TXN-{self.transaction_id_counter}"
        self.transaction_id_counter += 1
        
        total_price = credit_amount * price_per_credit
        
        transaction = {
            'transaction_id': transaction_id,
            'buyer_id': buyer_id,
            'buyer_name': self.companies[buyer_id]['name'],
            'seller_id': seller_id,
            'seller_name': self.companies[seller_id]['name'],
            'listing_id': listing_id,
            'credit_amount': credit_amount,
            'price_per_credit': price_per_credit,
            'total_price': total_price,
            'status': 'COMPLETED',
            'timestamp': time.time(),
            'blockchain_hash': None
        }
        
        # Update listing
        for listing in self.listings:
            if listing['listing_id'] == listing_id:
                listing['available_amount'] -= credit_amount
                if listing['available_amount'] == 0:
                    listing['status'] = 'SOLD_OUT'
                break
        
        # Update company stats
        self.companies[buyer_id]['credits_owned'] += credit_amount
        self.companies[buyer_id]['total_trades'] += 1
        
        self.companies[seller_id]['credits_sold'] += credit_amount
        self.companies[seller_id]['total_trades'] += 1
        
        # Record on blockchain
        last_block = self.blockchain.get_last_block()
        proof = last_block['proof'] + 1
        previous_hash = self.blockchain.hash(last_block)
        
        new_block = self.blockchain.create_block(proof, previous_hash)
        new_block['data'].append({
            'type': 'CARBON_CREDIT_TRADE',
            'transaction_id': transaction_id,
            'buyer': self.companies[buyer_id]['name'],
            'seller': self.companies[seller_id]['name'],
            'amount': credit_amount,
            'price': price_per_credit,
            'total_value': total_price,
            'timestamp': time.time()
        })
        
        transaction['blockchain_hash'] = self.blockchain.hash(new_block)
        
        self.transactions.append(transaction)
        
        return transaction
    
    def get_active_listings(self, filters=None):
        """Get all active marketplace listings with optional filters"""
        active = [l for l in self.listings if l['status'] == 'ACTIVE']
        
        if filters:
            if 'max_price' in filters:
                active = [l for l in active if l['price_per_credit'] <= filters['max_price']]
            if 'min_amount' in filters:
                active = [l for l in active if l['available_amount'] >= filters['min_amount']]
        
        return sorted(active, key=lambda x: x['created_at'], reverse=True)
    
    def get_transaction_history(self, company_id=None):
        """Get transaction history, optionally filtered by company"""
        if company_id:
            return [
                t for t in self.transactions
                if t['buyer_id'] == company_id or t['seller_id'] == company_id
            ]
        return self.transactions
    
    def get_market_stats(self):
        """Get overall marketplace statistics"""
        total_volume = sum(t['total_price'] for t in self.transactions)
        total_credits_traded = sum(t['credit_amount'] for t in self.transactions)
        
        avg_price = total_volume / total_credits_traded if total_credits_traded > 0 else 0
        
        return {
            'total_companies': len(self.companies),
            'active_listings': len([l for l in self.listings if l['status'] == 'ACTIVE']),
            'total_transactions': len(self.transactions),
            'total_volume_usd': round(total_volume, 2),
            'total_credits_traded': round(total_credits_traded, 2),
            'average_price_per_credit': round(avg_price, 2),
            'last_updated': time.time()
        }
    
    def get_company_profile(self, company_id):
        """Get detailed company profile"""
        if company_id not in self.companies:
            return None
        
        company = self.companies[company_id].copy()
        
        # Add transaction history
        company['transactions'] = self.get_transaction_history(company_id)
        company['active_listings'] = [
            l for l in self.listings 
            if l['seller_id'] == company_id and l['status'] == 'ACTIVE'
        ]
        
        return company
